x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/CompositeBucketsChangeCollector.java

```diff
@@ -492,20 +670,26 @@ public boolean processSearchResponse(final SearchResponse searchResponse) {
             return true;
         }
 
-        final CompositeAggregation agg = aggregations.get(compositeAggregation.getName());
-
-        Collection<? extends Bucket> buckets = agg.getBuckets();
-        afterKey = agg.afterKey();
+        boolean isDone = true;
 
-        if (buckets.isEmpty()) {
-            return true;
+        // collect changes from composite buckets
+        if (compositeAggregation != null) {
```

Comment:
Why are you doing this `null` check? Wouldn't it be better to rely on if any of `fieldCollector` items `requiresCompositeSources` ?

Reply:
yes, that's an alternative, however the null check is more robust regarding bugs.

Reply:
The only reason it is more robust in regarding bugs is because it is set to null explicitly earlier. 

I am not sure that is even necessary. It seems to me that some external thing is knowledgable of the internal logic of this system and is making changes. 

The separation of concerns seem to be bleeding across a bit.

Reply:
I moved the creation of the composite agg part into the change collector, that way I also addressed your concern regarding creating the composite but than throwing it away. The also made the checks a bit easier, although we still need some as you can have any combination of aggs, depending on which collectors are used.